File: backend/app/schemas/purchase.py
"""
Schemas de Compra y OCR
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime, date


# --- Items de compra ---

class PurchaseItemCreate(BaseModel):
    """Item para agregar a una compra"""
    product_id: int
    quantity: int = Field(..., gt=0)
    unit_cost: float = Field(..., gt=0)
    expiration_date: Optional[date] = None


class PurchaseItemResponse(BaseModel):
    """Respuesta de item de compra"""
    id: int
    product_id: int
    product_name: str = ""
    product_code: str = ""
    quantity: int
    unit_cost: float
    subtotal: float
    expiration_date: Optional[date]
    
    class Config:
        from_attributes = True


# --- Compra completa ---

class PurchaseCreate(BaseModel):
    """Schema para crear una compra manual"""
    supplier_id: Optional[int] = None
    items: List[PurchaseItemCreate] = Field(..., min_length=1)
    notes: Optional[str] = None
    purchase_date: Optional[date] = None


class PurchaseResponse(BaseModel):
    """Schema de respuesta para compra"""
    id: int
    supplier_id: Optional[int]
    supplier_name: Optional[str] = None
    total: float
    notes: Optional[str]
    receipt_image_path: Optional[str]
    user_id: int
    user_name: str = ""
    purchase_date: Optional[date]
    created_at: datetime
    items: List[PurchaseItemResponse] = []
    
    class Config:
        from_attributes = True


# --- OCR ---

class OCRProductMatch(BaseModel):
    """Producto detectado por OCR con su match en catálogo"""
    ocr_name: str  # Nombre extraído del ticket
    ocr_quantity: int
    ocr_unit_price: float
    ocr_subtotal: float
    
    # Match con catálogo
    matched_product_id: Optional[int] = None
    matched_product_name: Optional[str] = None
    match_confidence: float = 0  # 0-1
    status: str = "review"  # 'match', 'review', 'new'


class OCRResult(BaseModel):
    """Resultado del procesamiento OCR de un ticket"""
    supplier_name: Optional[str] = None
    supplier_cuit: Optional[str] = None
    ticket_date: Optional[date] = None
    ticket_number: Optional[str] = None
    items: List[OCRProductMatch] = []
    subtotal: Optional[float] = None
    tax: Optional[float] = None
    total: Optional[float] = None
    raw_text: str = ""  # Texto crudo extraído
    image_path: str = ""


class OCRPurchaseCreate(BaseModel):
    """Schema para crear compra desde OCR (después de revisión)"""
    ocr_result: OCRResult
    confirmed_items: List[PurchaseItemCreate]
    supplier_id: Optional[int] = None
    notes: Optional[str] = None

File: frontend/src/types/purchase.ts
/**
 * Types de Compra y OCR
 */

export interface PurchaseItem {
  id: number;
  product_id: number;
  product_name: string;
  product_code: string;
  quantity: number;
  unit_cost: number;
  subtotal: number;
  expiration_date?: string;
}

export interface Purchase {
  id: number;
  supplier_id?: number;
  supplier_name?: string;
  total: number;
  notes?: string;
  receipt_image_path?: string;
  user_id: number;
  user_name: string;
  purchase_date?: string;
  created_at: string;
  items: PurchaseItem[];
}

export interface PurchaseItemCreate {
  product_id: number;
  quantity: number;
  unit_cost: number;
  expiration_date?: string;
}

export interface PurchaseCreate {
  supplier_id?: number;
  items: PurchaseItemCreate[];
  notes?: string;
  purchase_date?: string;
}

// OCR Types
export interface OCRProductMatch {
  ocr_name: string;
  ocr_quantity: number;
  ocr_unit_price: number;
  ocr_subtotal: number;
  matched_product_id?: number;
  matched_product_name?: string;
  match_confidence: number;
  status: 'match' | 'review' | 'new';
}

export interface OCRResult {
  supplier_name?: string;
  supplier_cuit?: string;
  ticket_date?: string;
  ticket_number?: string;
  items: OCRProductMatch[];
  subtotal?: number;
  tax?: number;
  total?: number;
  raw_text: string;
  image_path: string;
}

export interface Supplier {
  id: number;
  name: string;
  cuit?: string;
  phone?: string;
  email?: string;
  address?: string;
  notes?: string;
  is_active: boolean;
  total_purchases: number;
  total_spent: number;
  created_at: string;
}
